### DISCUSSION TITLE
Can someone please help me understand sweeps?

### DISCUSSION THREAD

**Discussion by geiseri** [CadQuery/cadquery#1296] [Q&A]
I have a code [example](uhttps://gitlab.com/geiseri/cadquery_models/-/blob/2f4863124cc42c559a09ead855277930d024d7b1/top_groove.p) that is trying to do the following:

1. Select the top face of a shell that has been split.
2. Take that face's outer wire and offset a bit from the outer wire.
3. Sweep a sketch profiling the feature I want along that wire.

So what I am doing is:

1. Create the object and select the wire on line [19](https://gitlab.com/geiseri/cadquery_models/-/blob/2f4863124cc42c559a09ead855277930d024d7b1/top_groove.py#L19)
2. Do the offset on lines [36-47 ](https://gitlab.com/geiseri/cadquery_models/-/blob/2f4863124cc42c559a09ead855277930d024d7b1/top_groove.py#L36-47)I create a the of the sketch on line [10](https://gitlab.com/geiseri/cadquery_models/-/blob/2f4863124cc42c559a09ead855277930d024d7b1/top_groove.py#L10).

I end up with the following:

![image](https://user-images.githubusercontent.com/2265693/227803352-4c22fe36-969a-436f-9b40-f8583466f98b.png)

So questions: 

1. Why do I need to do the dance to create the sweep (is that the correct term for `plineSweep`?).  Is it because there are multiple edges in the wire?
2. I "think" I understand the workplace in "XZ" because it needs to be perpendicular to the path you are sweeping.  Why is the Z axis "off" from the original wire?  How do I calculate that offset to translate it back to the desired Z location.
3. The last and most confusing to me is that depending on the order of the edges (due to #1295) in the wire the final sweep is either offset differently on the Z axis, or it's just completely messed up. 
![image](https://user-images.githubusercontent.com/2265693/227803809-54a9368f-b2fc-4f16-bda5-7681fbf1e73c.png)

Now I am sure that this is due to my not really understanding what the sweep API is doing.  Could someone save me from my ignorance?  =)

**Comment by lorenzncode** [comment]
Please share minimal reproducible/working examples without extensions and other imports. I tried to make changes and reduce the code to the following: 

```py
import cadquery as cq

sk: cq.Sketch = (
    cq.Sketch()
    .arc((0, 0), 1.0, 0.0, 360.0)
    .arc((1, 1.5), 0.5, 0.0, 360.0)
    .segment((0.0, 2), (-1, 3.0))
    .hull()
)

box: cq.Workplane = cq.Workplane("XY").box(20, 30, 10).shell(2).split(keepBottom=True)

wire = box.faces(">Z").wires(cq.selectors.LengthNthSelector(-1))
wire2 = wire.toPending().offset2D(-0.25)

# I don't know which orientation was expected; you can also try "<XY" 
vert = wire2.vertices(">XY")

plineSweep = cq.Workplane("XZ", origin=vert.val().toTuple()).placeSketch(sk)
plineSweep = plineSweep.sweep(wire2)
plineSweep = plineSweep.translate((0, 0, 5))  # move up for visualization
```

>1. Why do I need to do the dance to create the sweep 

The original code calls `placeSketch` at all vertices.  Since the profile does not change this is unnecessary.  See the updated version. 


>2. ... Why is the Z axis "off" from the original wire? 

This is due to moving in the XZ plane:
```
        .moveTo(vert.X,vert.Y)
```

The origin was moved in the Z direction by `vert.Y`.


>3.  The last and most confusing to me is that depending on the order of the edges

Yes, the edges may be returned in any order.  It looks like it was offset differently based on the vert.Y value of the first vertex. 

This case doesn't require a specific edge/vertex ordering AFAICT.  Instead select a single point for the start of the sweep. 



I don't know if this is the exact result you expect.  You may need to make changes to the orientation of the sketch profile.

**Reply by geiseri** [reply]
Oh sorry, I will amend it to make sure.  I copy and paste the code from a larger project.   I did notice that if I did not provide those points it would not change the "direction" of the sketch.  That might have been because I was using the underlying wire and not the workplane context.  If I change to that I can see it works as intended per the Z-axis position.  I ended up going with the underlying shapes because I was having a hard time mapping the process I was doing in fusion360 to workplane concepts.  Of course this is because I have a very cursory understanding of CAD in the first place. :\

One follow-up for the desired wire selector.  Can confirm how `wire = box.faces(">Z").wires(cq.selectors.LengthNthSelector(-1))`  would decide it is the outer wire?  Is that because by nature of being the outside it would have to be longer? Then the `-1` would then cause it to index from the back of the list when sorted by length. 

That only leaves the path order for the sweep.   

I set the `vert` selection to `vert = wire2.vertices("<XY")` to get the shape I desire:
![image](https://user-images.githubusercontent.com/2265693/228010816-c9d75397-338a-40b6-a5f9-794d399b6b93.png)

But only in some cases.  I put the boxes back in to express the change in order. The colors change from red being the first point and blue as the last one in the list.  In some positions, the order changes and makes the sweep act curiously. 

![image](https://user-images.githubusercontent.com/2265693/228007335-3c5d3d7f-934d-49ec-bfd5-2cc1eb3a2de5.png)

![image](https://user-images.githubusercontent.com/2265693/228007882-0e78157d-bac0-49a8-88a6-1256c4224d47.png)

![image](https://user-images.githubusercontent.com/2265693/228008137-0afcaf8f-1a9f-4e80-a6b3-02804fd622b0.png)

In some cases the order works out as expected:

![image](https://user-images.githubusercontent.com/2265693/228008895-7a1927ca-ca49-4c77-9561-8ead28824eb4.png)

![image](https://user-images.githubusercontent.com/2265693/228009265-6b89cf2e-c8b5-4450-977c-c6338d2aa221.png)

Could it be that there are cases where the start and end of the sweep do not match up with the vertices it expects?  Again, I apologize, I am sure that this all stems from a very loose understanding of all of this.

**Reply by geiseri** [reply]
Sorry, after re-reading your response further, did you explain the 2nd part to me and I didn't get it?

**Reply by lorenzncode** [reply]
> Can confirm how wire = box.faces(">Z").wires(cq.selectors.LengthNthSelector(-1)) would decide it is the outer wire?


With `LengthNthSelector` you can select the nth object from the sorted list.  -1 selects the longest object due to Python list indexing where negative values are relative to the end of the sequence.


> Is that because by nature of being the outside it would have to be longer? Then the -1 would then cause it to index from the back of the list when sorted by length.

Yes. 


> But only in some cases.

My mistake.  We want to select a single vertex.  `wire2.vertices("<XY")` selected two.
Try:

```py
import cadquery as cq

sk: cq.Sketch = (
    cq.Sketch()
    .arc((0, 0), 1.0, 0.0, 360.0)
    .arc((1, 1.5), 0.5, 0.0, 360.0)
    .segment((0.0, 2), (-1, 3.0))
    .hull()
)

box: cq.Workplane = cq.Workplane("XY").box(20, 30, 10).shell(2).split(keepBottom=True)

wire = box.faces(">Z").wires(cq.selectors.LengthNthSelector(-1))
wire2 = wire.toPending().offset2D(-0.25)

vert = wire2.vertices("<X").vertices("<Y")

plineSweep = cq.Workplane("XZ", origin=vert.val().toTuple()).placeSketch(sk)
plineSweep = plineSweep.sweep(wire2)
plineSweep = plineSweep.translate((0, 0, 5))  # move up for visualization
```

**Reply by geiseri** [reply]
Yes, that did it, now with only the single vertex selected the sweep is correct no matter the order in the wire.

Now, one last clarification on this, and I think I have got it. 

For the `origin==vert.val().toTuple()`, is that telling the sweep to start at that point in the wire so it can walk the points back to where it started?

**Reply by lorenzncode** [reply]
The origin places the center of the profile on the sweep path.  The vertex was selected and plane "XZ" chosen based on the expected orientation of the profile. 

Say to create a solid ring with rectangular profile I might define the origin at any of these points:

```py
import cadquery as cq

sk = cq.Sketch().rect(5, 10)

path = cq.Workplane().circle(20)

result1 = cq.Workplane("XZ", origin=(-20, 0, 0)).placeSketch(sk).sweep(path)
result2 = cq.Workplane("XZ", origin=(20, 0, 0)).placeSketch(sk).sweep(path)
result3 = cq.Workplane("YZ", origin=(0, -20, 0)).placeSketch(sk).sweep(path)
result4 = cq.Workplane("YZ", origin=(0, 20, 0)).placeSketch(sk).sweep(path)
```
